src/main/java/uk/gov/hmcts/reform/pip/data/management/service/helpers/listmanipulation/CrownWarnedPddaListHelper.java

```diff
@@ -0,0 +1,127 @@
+
+        JsonNode warnedList = warnedPddaListData.get("WarnedList");
+        JsonNode courtLists = warnedList.get("CourtLists");
+
+        Optional.ofNullable(courtLists)
+            .filter(JsonNode::isArray)
+            .ifPresent(lists -> lists.forEach(courtList -> {
+                // Process WithFixedDate
+                Optional.ofNullable(courtList.get("WithFixedDate"))
+                    .filter(JsonNode::isArray)
+                    .ifPresent(withFixedDate -> withFixedDate.forEach(withFixDate ->
+                        formatFixture(withFixDate, groupedData, false)));
+
+                // Process WithoutFixedDate
+                Optional.ofNullable(courtList.get("WithoutFixedDate"))
+                    .filter(JsonNode::isArray)
+                    .ifPresent(withoutFixedDate -> withoutFixedDate.forEach(withoutFixDate ->
+                        formatFixture(withoutFixDate, groupedData, true)));
+            }));
+
+        // Sort cases by fixed date within each group
+        groupedData.forEach((key, cases) -> {
+            cases.sort(Comparator.comparing(CrownWarnedPddaList::getFixedDateAsLocalDate,
+                                            Comparator.nullsLast(Comparator.naturalOrder())));
+        });
+
+        return groupedData;
+    }
+
+    private static void formatFixture(JsonNode fixtureDate, Map<String,
+        List<CrownWarnedPddaList>> groupedData, boolean isWithoutFixedDate) {
+        JsonNode fixture = fixtureDate.get("Fixture");
+        String fixedDate = fixture.has("FixedDate") ? fixture.get("FixedDate").asText() : null;
+
+        Optional.ofNullable(fixture.get("Cases"))
+            .filter(JsonNode::isArray)
+            .ifPresent(cases -> cases.forEach(hearingCase -> {
+                Optional.ofNullable(hearingCase.get("Hearing"))
+                    .filter(JsonNode::isArray)
+                    .ifPresent(hearings -> hearings.forEach(hearing -> {
+                        String hearingDescription = isWithoutFixedDate
+                            ? "To be allocated"
+                            : hearing.has("HearingDescription") ? hearing.get("HearingDescription").asText() : "";
+
+                        groupedData.computeIfAbsent(hearingDescription, k -> new ArrayList<>())
+                            .add(formatCaseInformation(fixedDate, hearing, hearingCase));
+                    }));
+            }));
+    }
+
+    private static CrownWarnedPddaList formatCaseInformation(String fixedDate, JsonNode hearing,
+                                                             JsonNode hearingCase) {
+        String defendantNames = "";
+        if (hearingCase.has("Defendants")) {
+            defendantNames = CrownPddaListHelper.formatDefendantName(hearingCase.get("Defendants"));
+        }
+
+        String prosecutingAuthority = "";
+        if (hearingCase.has("Prosecution") && hearingCase.get("Prosecution").has("ProsecutingAuthority")) {
```

Comment:
Similar here, I think this can be replaced with:

prosecutingAuthority = findAndReturnNodeText(hearingCase.get("Prosecution"), "ProsecutingAuthority")